### PR TITLE
ReplayGenerator interactor implementation

### PR DESCRIPTION
**Pull request by hoodiexxx**
incomplete implementation, need to amend later. exactly implementation should depend on other use cases' implementation.

### CODE REVIEW

**Review by sbrad044 — COMMENTED** [review]
Good start. If this is duplicating the changes in Theo's pull request, are we allowed to approve both or do we have to change one of them?

**Comment by 6167656e74323431** [comment]
> If this is duplicating the changes in Theo's pull request, are we allowed to approve both or do we have to change one of them?

Under the hood, as it would turn out, the duplicated changes are the **same** changes, so as it turns out merging this is equivalent* to merging #42 and then this (but this is not true if we merge this and then #42).

*almost... you can look into git internals yourself, *if you dare*.

**Review by siddharthgowda — DISMISSED** [review]
This looks good only thing is that some methods will change based on the implementation of the other features. I know this was already mentioned for some of the methods but there are other methods where this also applies.

File: src/main/java/com/mg105/entities/Inventory.java [hunk 1]
```diff
@@ -83,7 +83,13 @@ public boolean removeItem(String itemName) {
...
         return false;
     }
-
+    public void removeAll()
+    {
+        for (int i = 0; i < numberOfItems(); i++)
+        {
+                this.items.remove(i);
+        }
+    }
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/entities/Inventory.java:92]: Style: this does not look like the automatic formatter was run.

File: src/main/java/com/mg105/use_cases/ReplayGenerator.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        inventory.removeAll();
+    }
+
+    public void Replay(GameState state, BattleCharacter battleCharacter)
+    {
+    // incomplete remake, need to amend later. exactly implementation should depend on other use cases' implementation
+        MapGenerator isekai = new MapGenerator(state);
+        BattleCharacter reincarnationPlayer = AttributeInheritance(battleCharacter);
+    }
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/use_cases/ReplayGenerator.java:28]: Style: this does not look like the automatic formatter was run.

**Review by 6167656e74323431 — CHANGES_REQUESTED** [review]
Correct me if I'm wrong, but it does not look like IDEA-formatted code.

See [this](https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html#reformat_module_directory).

File: src/main/java/com/mg105/entities/GameState.java [hunk 3]
```diff
@@ -1,7 +1,57 @@
...
+     * @param lastRoom  the end room of the map (where the game is won).
+     */
+    public void setMap(Room firstRoom, Room lastRoom) {
+        this.firstRoom = firstRoom;
+        this.currentRoom = firstRoom;
+        this.lastRoom = lastRoom;
+    }
 
+    /**
+     * Get the room the player happes to currently be in.
```
Comment (zhaoan12) [inline src/main/java/com/mg105/entities/GameState.java:26]: spelling error: change happes to happens

File: src/main/java/com/mg105/entities/Doorway.java [hunk 4]
```diff
@@ -1,7 +1,9 @@
 package com.mg105.entities;
 
+import java.awt.*;
+
 public class Doorway {
-    private final RoomPosition position;
+    private final Point position;
```
Comment (zhaoan12) [inline src/main/java/com/mg105/entities/Doorway.java:6]: I like how you gave a better name for the variable.

File: src/main/java/com/mg105/entities/GameState.java [hunk 5]
```diff
@@ -1,7 +1,57 @@
...
+    /**
+     * Get if the player is in the first room.
+     *
+     * @return true if the player is in the first room, false otherwise.
+     */
+    public boolean isCurrentRoomFirstRoom() {
+        // NOTE: In this case it is actually ok to directly compare the two via == and not .equals()... This is because
+        // we actually care that this is exactly the same instance of the room, not just two rooms that happen to have
+        // the same configuration.
+        return currentRoom == firstRoom;
```
Comment (zhaoan12) [inline src/main/java/com/mg105/entities/GameState.java:43]: Nice method, I will use it when implementing the tutorial.

**Review by zhaoan12 — APPROVED** [review]
Nice classes and methods. Since you did not give any test cases, maybe some ideas for the test case are to give the player extra stats in addition to their base stats. Then, use the replay generator to make a new game and see if the stats are retained.

File: src/main/java/com/mg105/use_cases/ReplayGenerator.java [hunk 6]
```diff
@@ -8,21 +8,35 @@ public class ReplayGenerator {
...
 
-    public BattleCharacter AttributeInheritance(BattleCharacter battleCharacter)
-    {
+    /**
+     * let a new BattleCharacter that inherit all the previous player's attributes
+     *
+     * @param battleCharacter the battleCharacter in the previous game
+     * @return a new BattleCharacter that inherit all the previous player's attributes
+     */
+    public BattleCharacter AttributeInheritance(BattleCharacter battleCharacter) {
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/use_cases/ReplayGenerator.java:17]: Style: we use `camelCase` for functions in Java... not `PascalCase`. Note: you do this in muptiple places but I will only highlight the first one I see.

**Review by 6167656e74323431 — CHANGES_REQUESTED** [review]

**Review by siddharthgowda — CHANGES_REQUESTED** [review]
Overall, it looks pretty good. The only thing is that there is no javadoc for the classes. Also make sure when you write the java doc that it is above the class.

**Comment by 6167656e74323431** [comment]
Question, what are the licensing terms of `imgs/gamelose.jpg`? Are we allowed to use it copyright-wise?

**Comment by 6167656e74323431** [comment]
Note: background2 i licensed under CC Attribution-NonCommercial-NoDerivs 2.0 Generic (CC BY-NC-ND 2.0)
Fred Seibert

File: src/main/java/com/mg105/interface_adapters/Toggler.java [hunk 7]
```diff
@@ -34,7 +34,8 @@ enum ToggleableComponent {
         BATTLE,
         /** The tutorial window */
         TUTORIAL,
-        /** The character selection screen. */
-        WALK_MENU
+        WALK_MENU,
+        LOSE_MENU
+
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/interface_adapters/Toggler.java:39]: Docstrings removed? Ad them back pls.

**Review by m-grigoriu — DISMISSED** [review]
Everything looks good.

File: src/main/java/com/mg105/user_interface/ReplayGeneratorButton.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package com.mg105.user_interface;
+
+import com.mg105.interface_adapters.ReplayGeneratorInterpreter;
+import javafx.event.ActionEvent;
+import org.jetbrains.annotations.NotNull;
+import com.mg105.interface_adapters.Toggler;
+import javafx.event.EventHandler;
+
+import javax.swing.*;
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/user_interface/ReplayGeneratorButton.java:9]: This import should be removed (we don't use swing afaik)

**Review by 6167656e74323431 — CHANGES_REQUESTED** [review]

File: src/main/java/com/mg105/use_cases/ReplayGenerator.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * @param battleCharacter the battleCharacter in the previous game
+     * @return a new BattleCharacter that inherit all the previous player's attributes
+     */
+
+    public ReplayGenerator(GameState state) {
+        this.state = state;
+    }
+
+    public ArrayList<BattleCharacter> attributeInheritance() {
```
Comment (zhaoan12) [inline src/main/java/com/mg105/use_cases/ReplayGenerator.java:26]: need doctring i think

**Review by zhaoan12 — APPROVED** [review]
Nice class

**Review by 6167656e74323431 — CHANGES_REQUESTED** [review]
See my earlier comments

**Review by 6167656e74323431 — APPROVED** [review]

**Review by zhaoan12 — APPROVED** [review]
Nice changes

**Review by m-grigoriu — APPROVED** [review]

**Pull request merged**